src/libasr/pass/intrinsic_functions.h

```diff
@@ -1527,6 +1528,42 @@ namespace Btest {
 
 } // namespace Btest
 
+namespace Ibits {
+
+    static ASR::expr_t *eval_Ibits(Allocator &al, const Location &loc,
+            ASR::ttype_t* t1, Vec<ASR::expr_t*> &args, diag::Diagnostics& /*diag*/) {
+        int64_t val1 = ASR::down_cast<ASR::IntegerConstant_t>(args[0])->m_n;
+        int64_t val2 = ASR::down_cast<ASR::IntegerConstant_t>(args[1])->m_n;
+        int64_t val3 = ASR::down_cast<ASR::IntegerConstant_t>(args[2])->m_n;
+        int64_t result;
+        result = (val1 >> val2) & ((1 << val3) - 1);
+        return make_ConstantWithType(make_IntegerConstant_t, result, t1, loc);
+    }
+
+    static inline ASR::expr_t* instantiate_Ibits(Allocator &al, const Location &loc,
+            SymbolTable *scope, Vec<ASR::ttype_t*>& arg_types, ASR::ttype_t *return_type,
+            Vec<ASR::call_arg_t>& new_args, int64_t /*overload_id*/) {
+        declare_basic_variables("");
+        fill_func_arg("x", arg_types[0]);
+        fill_func_arg("y", arg_types[1]);
+        fill_func_arg("z", arg_types[2]);
+        auto result = declare(fn_name, return_type, ReturnVar);
+        
+        ASR::expr_t *val1 = args[0];
+        ASR::expr_t *val2 = args[1];
+        ASR::expr_t *val3 = args[2];
+        ASR::expr_t *one = i(1, arg_types[0]);
+
+        body.push_back(al, b.Assignment(result, i_BitAnd(i_BitRshift(val1, val2, return_type), iSub(i_BitLshift(one, val3, return_type), one), return_type)));
```

Comment:
The problem is when the kinds of the arguments differ. The following will throw error.
```
program main
    print*, ibits(20_8, 2_4, 2_4) 
end
```
According to the definition, the kind of result will be equal to the kind of first argument. So I guess we can cast the 2nd and 3rd argument to the kind of 1st argument.

Use the following:
```suggestion
        ASR::expr_t *cast_val2 = ASRUtils::EXPR(ASR::make_Cast_t(al, loc, val2, ASR::cast_kindType::IntegerToInteger, arg_types[0], nullptr));
        ASR::expr_t *cast_val3 = ASRUtils::EXPR(ASR::make_Cast_t(al, loc, val3, ASR::cast_kindType::IntegerToInteger, arg_types[0], nullptr));
        body.push_back(al, b.Assignment(result, i_BitAnd(i_BitRshift(val1, cast_val2, return_type), iSub(i_BitLshift(one, cast_val3, return_type), one), return_type)));
```

Reply:
Then this should work fine.